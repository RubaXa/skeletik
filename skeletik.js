(function (factory) {
	'use strict';

	if (typeof module === 'object' && module.exports) {
		module.exports = factory();
	} else if (typeof define === 'function' && define.amd) {
		define([], factory);
	} else {
		window.skeletik = factory();
	}
})(function () {
	'use strict';

	/**
	 * @class Lexer
	 * @param {string} input
	 */
	function Lexer(input) {
		this.line = 1;
		this.column = 1;
		this.idx = 0;
		this.lastIdx = 0;
		this.indent = {tab: 0, space: 0};
		this.code = null;
		this.prevCode = null;
		this.input = input;
		this.length = input.length;
		this.skipLength = 0;
	}

	Lexer.prototype = /** @lends Lexer# */{
		constrcutor: Lexer,

		getToken: function (leftOffset, rightOffset) {
			return this.input.substring(this.lastIdx + (leftOffset|0), this.idx + (rightOffset|0));
		},

		save: function () {
			this.lastIdx = this.idx + 1;
			return this;
		},

		takeToken: function (leftOffset, rightOffset) {
			var token = this.getToken(leftOffset, rightOffset);

			this.lastIdx = this.idx;
			this.taked = true;

			return token;
		},

		getChar: function () {
			return String.fromCharCode(this.code);
		},

		takeChar: function () {
			this.lastIdx = this.idx;
			return String.fromCharCode(this.code); // todo: vs. charAt()
		},

		peek: function (offset) {
			return offset === 0 ? this.code : this.input.charCodeAt(this.idx + offset);
		},

		error: function (message, bone, columnOffset) {
			var error = new SyntaxError(message);

			error.bone = bone;
			error.chr = this.getChar();
			error.token = this.getToken();
			error.line = this.line;
			error.column = this.column + (columnOffset|0);
			error.details = this.input.split('\n')[error.line - 1].substr(0, error.column);
			error.pretty = [
				error.details.replace(/\t/g, ' '),
				new Array(error.column).join('-') + '^'
			].join('\n');

			throw error;
		}
	};


	/**
	 * @class Bone
	 * @param {string} type
	 * @param {*} [raw]
	 * @constructor
	 */
	function Bone(type, raw) {
		this.type = type;

		if (raw !== void 0) {
			this.raw = raw;
		}

		this.length = 0;
		this.nodes = [];
	}

	Bone.prototype = /** @lends Bone# */{
		constrcutor: Bone,

		add: function (type, raw) {
			var bone = new Bone(type, raw);

			bone.parent = this;

			if (this.length > 0) {
				bone.prev = this.last;
				this.last.next = bone;
			}

			this.length = this.nodes.push(bone);
			this.first = this.nodes[0];
			this.last = this.nodes[this.length - 1];

			return this;
		},

		toJSON: function () {
			return {
				type: this.type,
				raw: this.raw,
				nodes: this.length ? this.nodes : [] // todo: Надо поправить на всегда массив
			};
		}
	};


	/**
	 * @class Skeletik;
	 * @param ranges
	 * @param spec
	 * @constructor
	 */
	function Skeletik(ranges, spec) {
		this.ranges(ranges);
		this.spec(spec);

		return this.compile();
	}

	Skeletik.prototype = /** @lends Skeletik# */{
		constrcutor: Skeletik,
		Bone: Bone,

		ranges: function (ranges) {
			// Компилируем диапозоны
			Object.keys(ranges).forEach(function (name) {
				if (/[^$a-z_]/.test(name)) {
					throw new Error("skeletik#ranges: 'name' of range can consist only 'a-z', '_' or '$'");
				}

				ranges[name] = new Function('code', 'return (' + ranges[name].map(function (range) {
					if (range.length == 3) {
						return range.charCodeAt(0) + ' <= code && ' + range.charCodeAt(2) + ' >= code';
					} else {
						return range.charCodeAt(0) + ' === code';
					}
				}).join(') || (') + ')');
			});

			this._ranges = ranges;
		},

		spec: function (spec) {
			this._spec = spec;
			this._states = [];
			this._vars = [];
			this._events = {};

			Object.keys(spec).forEach(function (state) {
				var stateName = 'state_' + state.replace(/[^a-z0-9]/gi, '_');
				var rules = spec[state];
				var events = spec[state].__events;

				this._vars.push(stateName + ' = "' + state + '"');
				this._states[state] = stateName;
				this._events[state] = events;

				delete spec[state].__events;

				this._states.push({
					name: stateName,
					events: !!events,
					rules: Object.keys(rules).map(function (chr) {
						var next = rules[chr];
						var setLastIdx = true;
						var mode = typeof next === 'string' && next.charAt(0);

						if (mode === '!' || mode === '>') {
							next = next.substr(1);
							setLastIdx = mode;
						}

						return {
							chr: chr,
							code: chr.charCodeAt(0),
							next: next,
							setLastIdx: setLastIdx, 
						};
					})
				});
			}, this);
		},

		compile: function () {
			var _this = this;
			var code = (function (lex, bone, options) {
				var code = 0;
				var calcIndent = true;
				var state = '';
				var _state = '';
				var length = lex.length;
				var exit = false;
				var setLastIdx; // todo: rename to `cursorMode`

				this.vars();

				function doIt(chr) {
					var result = spec[state][chr](lex, bone);

					if (result == null) {
						state = '';
					} else {
						if (result === '->' || result === '-->') {
							setLastIdx = (result === '-->');
							return;
						} else if (result.length === 2 && !result.type) {
							bone = result[0];
							state = result[1];
						} else if (result.type) {
							bone = result;
							state = '';
							return;
						} else if (typeof result === 'string') {
							state = result;
						}

						if (state !== _state) {
							var mode = state.charAt(0); // "+" todo: charCodeAt
							if (mode === '!' || mode === '>') {
								state = state.substr(1);
								setLastIdx = mode;
							}
						}
					}
				}

				function emit(name) {
					var fn = events[state];

					if ((fn !== void 0) && (fn = fn[name]) !== void 0) {
						var result = fn(lex, bone);

						if (result === '->') {
							return false;
						} else if (result === false) {
							state = '';
							return false;
						}
					}
				}

				while (lex.idx < length) {
					code = lex.input.charCodeAt(lex.idx);

					lex.prevCode = lex.code;
					lex.code = code;
					lex.state = state;
					setLastIdx = false;

					if (options.onpeek !== void 0 && (options.onpeek(lex, bone) === false)) {
						lex.code = code = 10;
						exit = true;
					}

					if (calcIndent) {
						if (code === 9 || code === 32) {
							lex.indent[code === 9 ? 'tab' : 'space']++;
						} else {
							calcIndent = false;
							var retVal = (options.onindent !== void 0) && options.onindent(lex, bone);

							if (retVal !== void 0 && retVal.type) {
								bone = retVal;
							}
						}
					}

					this.loop(state, code);

					if (exit) {
						break;
					}

					if (code === 10 && (length - lex.idx) !== 1) {
						emit("line");
						lex.line++;
						lex.column = 0;
						lex.indent = {tab: 0, space: 0};
						calcIndent = true;
					}

					if (state !== _state) {
						emit("start");
						_state = state;
					}

					if (setLastIdx && setLastIdx !== '>') {
						if (setLastIdx === '!') {
							lex.lastIdx = lex.idx;
						} else {
							lex.lastIdx = lex.idx + 1;
						}
					}
					
					if (setLastIdx !== '>') {
						lex.idx++;
						lex.column++;
					}
				}

				return bone;
			}).toString().replace(/([ \t]*)this\.(.*?)\((?:(\w+),\s*(\w+))?\);/g, function (_, spaces, name, vState, vCode) {
				if (name === 'vars') {
					return spaces + 'var ' + _this._vars.join(';\n' + spaces + 'var ');
				}
				else if (name === 'loop') {
					return spaces + _this._states.map(function (state) {
						var code = [];
						var rules = state.rules;

						code.push('\t' + rules.map(function (rule) {
							var code = [];
							var next = rule.next;
							var chr = JSON.stringify(rule.chr);

							if (_this._ranges[rule.chr]) {
								code.push('if (ranges.' + rule.chr + '(code)) { // char: ' + chr);
							} else {
								code.push(rule.chr !== '' ? 'if (' + rule.code + ' === code) { // char: ' + chr : '{ // char: any');
							}

							if (next === '->' || next === '-->') {
								code.push('\tstate = _state;');
								(state === '-->') && code.push('\tsetLastIdx = true;');
							} else {
								if (rule.setLastIdx) {
									code.push('\tsetLastIdx = ' + JSON.stringify(rule.setLastIdx) + ';');
								}

								if (typeof next === 'function') {
									code.push('\tdoIt(' + chr + ');');
								} else {
									code.push('\tstate = ' + _this._states[next] + ';');
								}
							}

							code.push('}');

							return code.join('\n' + spaces + '\t');
						}).join(' else '));

						if (state.events){
							code.unshift('if (emit("char") !== false) {');
							code.push('}');
						}

						return [].concat('if (' + state.name + ' === ' + vState + ') {', code, '}').join('\n' + spaces);
					}).join(' else ');
				}

				return '';
			});

			return new Function('ranges, spec, events', 'return ' + code)(this._ranges, this._spec, this._events);
		}
	};

	var skeletik = function (ranges, spec, options) {
		options = options || {};

		var parser = new Skeletik(ranges, spec);
		var parse = function (input) {
			var lex = new Lexer(input + '\n');
			var root = new Bone('#root');

			options.onstart && options.onstart(lex, root);

			var bone = parser(lex, root, options);
			var retEnd = options.onend && options.onend(lex, bone);

			if (retEnd && retEnd.type) {
				bone = retEnd;
			}

			return bone;
		};

		parse.capture = function (lex, localOptions) {
			localOptions = localOptions || {};

			var root = new Bone('#root');

			for (var key in options) {
				if (localOptions[key]) {
					localOptions[key] = (function (parent, local) {
						return function () {
							var parentVal = parent.apply(this, arguments);
							var localVal = local.apply(this, arguments);
							return localVal === void 0 ? parentVal : localVal;
						};
					})(options[key], localOptions[key])
				} else {
					localOptions[key] = options[key];
				}
			}

			localOptions.onstart && localOptions.onstart(lex, root);
			var bone = parser(lex, root, localOptions);
			localOptions.onend && localOptions.onend(lex, bone);

			return bone;
		};

		return parse;
	};

	// Export
	skeletik.Bone = Bone;
	skeletik.preset = {};
	skeletik.version = '0.1.0';

	return skeletik;
});
