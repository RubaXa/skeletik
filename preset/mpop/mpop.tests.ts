import { Loc, mpop } from './mpop';
import { existsSync, readFileSync } from 'fs';
import { Bone } from '../../src/skeletik/skeletik';

export const parse = (name: string, path = __dirname) => {
	const filename = `${path}/__fixture__/${name}.tpl`;

	// if (local && !existsSync(filename)) {
	// 	return false;
	// }

	const raw = readFileSync(filename) + '';

	// `JSON.stringify` + `JSON.parse needs` for converting bone to plain objects
	return {
		ast: JSON.parse(JSON.stringify(mpop(raw))),
		raw,
		filename,
	};
};

export const stringify = (bone: Bone) => {
	if (!bone) {
		return '';
	}

	const {
		type,
		raw,
		nodes,
	} = bone;

	if (type === '#root') {
		return nodes.map(stringify).join('');
	} else if (type === '#text') {
		return raw.value;
	} else if (type === 'VALUE') {
		return `##${raw.name}##`;
	} else if (type === 'CALL') {
		return `##${raw.name}(${raw.args.join(',')})##`;
	} else if (type === 'SET') {
		return `##SetVars(${raw.name}=${raw.value})##`;
	} else if (type === 'INCLUDE' || type === 'CONTINUE' || type === 'BREAK') {
		return `<!-- ${type}${raw.value ? ` ${raw.value}` : ''} -->`;
	} else if (type === 'FOR') {
		return `<!-- FOR ${raw.data} -->${nodes.map(stringify).join('')}<!-- /FOR -->`;
	} else if (/^(ELSE|IF)/.test(type)) {
		const {
			ending,
			consequent,
			alternate,
		} = raw;

		return `<!-- ${type}${raw.test ? ` ${raw.test}` : ''} -->${
			nodes.length ? nodes.map(stringify).join('') :
			[
				consequent ? consequent.nodes.map(stringify).join('') : '',
				stringify(alternate),
			].join('')
		}${/^IF/.test(type) ? `<!-- /${type}${ending ? ` ${ending}` : ''} -->` : ''}`;
	}
};

export const stringifyTypes = (bone: Bone) => {
	if (!bone) {
		return '';
	}

	const {
		type,
		raw,
		nodes,
	} = bone;

	if (type === '#root') {
		return nodes.map(stringifyTypes).join('');
	} else if (type === '#text') {
		return raw.value;
	} else if (['VALUE', 'CALL', 'SET', 'INCLUDE', 'CONTINUE', 'BREAK'].includes(type)) {
		return `[${type}]`;
	} else if (['FOR'].includes(type)) {
		return `[${type}]${nodes.map(stringifyTypes).join('')}[/${type}]`;
	} else if (/^(ELSE|IF)/.test(type)) {
		const {
			consequent,
			alternate,
		} = raw;

		return `[${type}]${
			nodes.length ? nodes.map(stringifyTypes).join('') :
			[
				consequent ? consequent.nodes.map(stringifyTypes).join('') : '',
				stringifyTypes(alternate),
			].join('')
		}${/^IF/.test(type) ? `[/${type}]` : ''}`;
	}
};

it('text', () => {
	const {ast} = parse('text');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('SetVars', () => {
	const {ast} = parse('set-vars');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('SetVars with expression', () => {
	const {ast} = parse('set-vars.expr');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('call', () => {
	const {ast} = parse('call');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('include', () => {
	const {ast} = parse('include');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('value', () => {
	const {ast} = parse('value');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('if', () => {
	const {ast} = parse('if');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('if-ending', () => {
	const {ast} = parse('if.ending');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('ifdef', () => {
	const {ast} = parse('if.def');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('ifnot', () => {
	const {ast} = parse('if.not');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('ifnotdef', () => {
	const {ast} = parse('if.not.def');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('if-else', () => {
	const {ast} = parse('if.else');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('if-elseif-else', () => {
	const {ast} = parse('if.elseif.else');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('if-elseif-notdef-else', () => {
	const {ast} = parse('if.elseif.notdef.else');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('if-else-if-else', () => {
	const {ast} = parse('if.else-if.else');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('if-else-if-notdef-else', () => {
	const {ast} = parse('if.else-if.notdef.else');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});

it('for', () => {
	const {ast} = parse('for');
	expect(stringify(ast)).toMatchSnapshot();
	expect(stringifyTypes(ast)).toMatchSnapshot();
});
