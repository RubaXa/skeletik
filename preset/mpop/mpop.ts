import {skeletikFactory as skeletik, charCode, Bone, Lexer} from '../../src/skeletik/skeletik';

export type Pos = [number, number];
export type Loc = {start: Pos; end: Pos};

function createSeq(seq: string, nextState: string, strict = false) {
	let cursor = 0;
	let length = seq.length;

	return (lex: Lexer) => {
		if (lex.code === seq.charCodeAt(cursor)) {
			cursor++;

			if (cursor === length) {
				cursor = 0;
				return nextState;
			} else {
				return '->';
			}
		}

		cursor = 0;
		return strict ? '->' : '';
	};
}

function fail(lex: Lexer, bone: Bone) {
	lex.error(`Invalid character \`${lex.getChar()}\`, state: ${lex.state}`, bone);
}

function getPos(lex: Lexer, offset = 0): Pos {
	return [lex.line, lex.column + offset];
}

const KEYWORDS: {[name: string]: boolean | string} = {
	'IF': true,
	'IFDEF': true,
	'IFNOT': true,
	'IFNOTDEF': true,
	'ELSE': true,
	'ELSEIF': true,
	'ELSEIFDEF': true,
	'ELSEIFNOT': true,
	'ELSEIFNOTDEF': true,
	'INCLUDE': 'shorty',
	'FOR': true,
	'CONTINUE': 'shorty',
	'BREAK': 'shorty',
};

const OPERANDS = {
	[charCode('&')]: 1,
	[charCode('|')]: 1,
};

const PARENTHESES = {
	[charCode('(')]: 1,
	[charCode(')')]: -1,
};

const C_LT = charCode('<');
const C_GT = charCode('>');
const C_HASH = charCode('#');
const C_ACCENT = charCode('!');
const C_MINUS = charCode('-');
const C_SPACE = charCode(' ');

let expr = '';
let isClose = false;
let openParentheses = 0;
let startPos: Pos;

function addText(lex: Lexer, bone: Bone, trim?: boolean) {
	const token = lex.takeToken();
	const len = token.length;

	if (!len) {
		return;
	}

	if (bone.last && bone.last.type === '#text') {
		bone.last.raw.value += token;
		// bone.last.raw.loc.end = getPos(lex, 0);
	} else {
		bone.add(new Bone('#text', {
			// loc: {start: getPos(lex, -len + 1), end: null},
			value: token,
		}));
	}

	return bone.last;
}


export const mpop = skeletik({
	'$ws': [' '],
	'$name': ['A-Z', 'a-z', '_', '0-9', '/'],
	'$cmd': ['A-Z'],
	'$start_expr': ['!', 'A-Z', 'a-z', '0-9'],
	'$expr': ['A-Z', 'a-z', '_', '0-9', '!'],
}, {
	'': {
		// __events: {
		// 	leave: (lex) => {
		// 		if (lex.code === C_LT || lex.code === C_HASH) {
		// 			startPos = getPos(lex);
		// 		}
		// 	},
		// },

		'<': (lex, bone) => {
			// <`!--`
			if (lex.peek(+1) === C_ACCENT && lex.peek(+2) === C_MINUS && lex.peek(+3)) {
				addText(lex, bone);
				lex.skipNext(3);
				return 'START_COMMENT';
			} else {
				return '->';
			}
		},

		'#': (lex, bone) => {
			if (lex.peek(+1) === C_HASH && lex.range.$expr(lex.peek(+2))) {
				addText(lex, bone);
				lex.skipNext(1);
				return 'AWAIT_FUNC_OR_VALUE';
			} else {
				return '->';
			}
		},

		'': '->',
	},

	'AWAIT_FUNC_OR_VALUE': {
		'$name': '->',
		'(': (lex, bone) => {
			const token = lex.getToken();

			if (token === 'SetVars') {
				return 'SET_VARS:NAME';
			} else {
				return [bone.add('CALL', {
					// loc: {
					// 	start: startPos,
					// 	end: null,
					// },
					name: token,
					args: [],
				}).last, 'FUNC:ARGS'];
			}
		},
		'#': (lex, bone) => {
			if (lex.peek(+1) === C_HASH) {
				bone.add('VALUE', {
					// loc: {
					// 	start: startPos,
					// 	end: getPos(lex, 2),
					// },
					name: lex.getToken(),
				});
				lex.skipNext(1);
				return '';
			}

			return fail(lex, bone);
		},
		'': fail,
	},

	'START_COMMENT': {
		'$ws': '->', // skip white spaces
		'/': () => {
			isClose = true;
			return 'AWAIT_CMD';
		},
		'$cmd': '!AWAIT_CMD',
		'': '',
	},

	'AWAIT_CMD': {
		'$cmd': '->', // continue
		' '(lex: Lexer, bone: Bone) {
			const token = lex.getToken().replace(' ', '');
			const isLikeIF = /^IF/.test(token);
			const isLikeElse = /^ELSE/.test(token);
			const kwType = KEYWORDS[token]

			if (kwType) {
				if (isClose) {
					isClose = false;

					if (isLikeIF) {
						bone = bone.parent;
					}

					while (/^ELSE/.test(bone.type)) {
						bone = bone.parent;
					}

					if (bone.type !== token) {
						lex.error(`Wrong closed tag "${token}", but must be ${bone.type}`);
					} else {
						return [bone, 'KW_AWAIT_END'];
					}
				}

				if (isLikeElse) {
					bone = bone.parent;
					bone.raw.alternate = new Bone(token, {});
					bone.raw.alternate.parent = bone;

					return [bone.raw.alternate, 'KW_ELSE'];
				}

				return [
					bone.add(token, {
						// loc: {start: startPos, end: null},
					}).last,
					isLikeIF ? 'KW_IF' : (kwType === true ? `KW_${token}` : `KW_TYPE_${kwType.toUpperCase()}`),
				];
			}
		},
		'': '',
	},

	'SET_VARS:NAME': {
		'=': (lex, bone) => {
			return [bone.add('SET', {
				// loc: {start: startPos, end: startPos},
				name: lex.getToken(),
			}).last, 'SET_VARS:VALUE'];
		},
		'': '->',
	},

	'SET_VARS:VALUE': {
		')': (lex, bone) => {
			bone.raw.value = lex.getToken();
			// bone.raw.loc.end = getPos(lex, 3);
			lex.skipNext(2);
			return bone.parent;
		},
		'': '->',
	},

	'FUNC:ARGS': {
		')': (lex, bone) => {
			bone.raw.args = [lex.getToken()];
			// bone.raw.loc.end = getPos(lex, 3);
			lex.skipNext(2);
			return bone.parent;
		},
		'': '->',
	},

	'KW_FOR': {
		'$start_expr': '!EXPR',
		'': fail,
	},

	'KW_FOR_EXPR': {
		'': (_, bone: Bone) => {
			bone.raw.data = expr;
		},
	},

	'KW_IF': {
		'$start_expr': '!EXPR',
		'': fail,
	},

	'KW_IF_EXPR': {
		'': (_, bone: Bone) => {
			bone.raw.test = expr;
			bone.raw.consequent = new Bone('#block');
			bone.raw.consequent.parent = bone;
			bone.raw.alternate = null
			return bone.raw.consequent;
		},
	},

	'KW_TYPE_SHORTY': {
		'': createSeq('-->', '>KW_TYPE_SHORTY_END', true),
	},

	'KW_TYPE_SHORTY_END': {
		'>': (lex, bone) => {
			const token = lex.getToken(0, -2).trim();
			bone.raw.value = token;
			// bone.raw.loc.end = getPos(lex, 1);
			return bone.parent;
		},
	},

	'KW_ELSE': {
		'$start_expr': (lex, bone) => {
			const token = lex.getToken();

			if (/^ *IF(NOT)?(DEF)? $/.test(token)) {
				bone.type += ` ${token.trim()}`;
				return '>KW_IF';
			} else if (/^ELSEIF/.test(bone.type)) {
				expr = '';
				return '>KW_IF';
			} else if ('IFNOTDEF'.substr(0, token.length) === token || 'IFDEF'.substr(0, token.length) === token) {
				return '->'
			} else {
				console.log(token);
				return 'fail';
			}
		},
		'-': (lex) => {
			if (lex.peek(+1) === C_MINUS && lex.peek(+2) === C_GT) {
				lex.skipNext(2);
				return '';
			}
		},
	},

	'KW_AWAIT_END': {
		'-': (lex, bone) => {
			if (lex.peek(+1) === C_MINUS && lex.peek(+2) === C_GT) {
				const token = lex.getToken().trim();

				token && (bone.raw.ending = token);
				lex.skipNext(2);
				// bone.raw.loc.end = getPos(lex, 1);
				return bone.parent;
			}
		},
	},

	'EXPR': {
		'$expr': '->',
		' ': '->',
		'': (lex: Lexer, bone: Bone) => {
			const {code} = lex;

			if (PARENTHESES[code]) {
				openParentheses += PARENTHESES[code];

				if (openParentheses < 0) {
					lex.error('Oops, unexpected close parenthesis.');
				}

				return '->';
			}

			if (openParentheses > 0 || OPERANDS[code]) {
				return '->';
			}

			if (code === C_MINUS && lex.peek(+1) === C_MINUS && lex.peek(+2) === C_GT) {
				expr = lex.getToken().trim();
				lex.skipNext(1);
				return `${lex.prevState}_EXPR`;
			}

			return fail(lex, bone);
		},
	},
}, {
	onend(lex: Lexer, bone: Bone, root) {
		if (bone !== root) {
			lex.error(`Parsing failed, "${bone.type}" is not closed, state: "${lex.prevState}" -> "${lex.state}"`);
		}

		const text = addText(lex, bone);
		text.raw.value = text.raw.value.replace(/[\r\n]+$/, '\n');
	}
});