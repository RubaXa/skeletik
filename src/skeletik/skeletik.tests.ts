import skeletik from './skeletik';

it('ranges:fail', () => {
	try {
		skeletik({'-': ['1']}, {});
	} catch (err) {
		expect(err.message).toBe("skeletik#ranges: 'name' of range can consist only 'a-z', '_' or '$'");
	}
});

it('a + b', () => {
	const parser = skeletik({
		'$e': ['a', 'b']
	}, {
		'': {
			'$e': '!expr',
			'+': function (lex, bone) {
				bone.add('#s', lex.takeChar());
			}
		},
		'expr': {
			'': function (lex, bone) {
				bone.add('#e', lex.takeToken());
				return '>';
			}
		}
	});

	expect(JSON.stringify(parser('a + b'), null, 2)).toBe(JSON.stringify({
		type: '#root',
		raw: null,
		nodes: [
			{type: '#e', raw: 'a', nodes: []},
			{type: '#s', raw: '+', nodes: []},
			{type: '#e', raw: 'b', nodes: []}
		]
	}, null, 2));
});
