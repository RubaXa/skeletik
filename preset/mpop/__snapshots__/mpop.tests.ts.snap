// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetVars 1`] = `
"##SetVars(TRUE=1)##
##SetVars(HOST=mail.ru)##
"
`;

exports[`SetVars 2`] = `
"[SET]
[SET]
"
`;

exports[`SetVars with expression 1`] = `
"##SetVars(VID=##GET_VID##)##
##SetVars(FOO_JIGURDA=FMAIL-123-##VID##)##
##SetVars(BAR_JIGURDA=##VID##-FMAIL-456)##
##SetVars(QUX_JIGURDA=feature-##IP##-##VID##-FMAIL-789)##
"
`;

exports[`SetVars with expression 2`] = `
"[SET]
[SET]
[SET]
[SET]
"
`;

exports[`call 1`] = `
"##Reklama/ShowBanner(5588)##
"
`;

exports[`call 2`] = `
"[CALL]
"
`;

exports[`for 1`] = `
"##SetVars(ARR_Items=2)##
##SetVars(ARR_1_a=1)##
##SetVars(ARR_2_a=2)##

<!-- FOR ARR -->
	...
	##.field1##
	##.field2##
	...
	<!-- IF EXPR1 --><!-- CONTINUE --><!-- /IF -->
	...
	<!-- IF EXPR2 --><!-- BREAK --><!-- /IF -->
	...
<!-- /FOR -->
"
`;

exports[`for 2`] = `
"[SET]
[SET]
[SET]

[FOR]
	...
	##.field1##
	##.field2##
	...
	[IF][CONTINUE][/IF]
	...
	[IF][BREAK][/IF]
	...
[/FOR]
"
`;

exports[`if 1`] = `
"<!-- IF !true -->
	Hi, ##UserName##
<!-- /IF -->
<!-- IF 0 -->Text<!-- /IF -->
"
`;

exports[`if 2`] = `
"[IF]
	Hi, [VALUE]
[/IF]
[IF]Text[/IF]
"
`;

exports[`if-else 1`] = `
"<!-- IF TestServer && Eq(GET_x,1) -->
	Hi, ##UserName##
<!-- ELSE -->
	##JsonEncode(GET_id)##
<!-- /IF -->
"
`;

exports[`if-else 2`] = `
"[IF]
	Hi, [VALUE]
[ELSE]
	[CALL]
[/IF]
"
`;

exports[`if-else-if-else 1`] = `
"<!-- IF TestServer && Eq(GET_x,1) -->
	Hi, ##UserName##!!
<!-- ELSE IF !!TestServer -->
	##HOST##
<!-- ELSE -->
	##JsonEncode(GET_id)##
<!-- /IF -->
"
`;

exports[`if-else-if-else 2`] = `
"[IF]
	Hi, [VALUE]!!
[ELSE IF]
	[VALUE]
[ELSE]
	[CALL]
[/IF]
"
`;

exports[`if-else-if-notdef-else 1`] = `
"<!-- IF TestServer && Eq(GET_x,1) -->
	Hi, ##UserName##!!
<!-- ELSE IFNOT GET_id -->
	IFNOT
<!-- ELSE IFDEF GET_email -->
	IFDEF
<!-- ELSE IFNOTDEF TestServer -->
	IFNOTDEF
<!-- ELSE -->
	##JsonEncode(GET_id)##
<!-- /IF -->
"
`;

exports[`if-else-if-notdef-else 2`] = `
"[IF]
	Hi, [VALUE]!!
[ELSE IFNOT]
	IFNOT
[ELSE IFDEF]
	IFDEF
[ELSE IFNOTDEF]
	IFNOTDEF
[ELSE]
	[CALL]
[/IF]
"
`;

exports[`if-elseif-else 1`] = `
"<!-- IF TestServer && Eq(GET_x,1) -->
	Hi, ##UserName##?!
<!-- ELSEIF !TestServer -->
	##HOST##
<!-- ELSE -->
	##JsonEncode(GET_id)##
<!-- /IF -->
"
`;

exports[`if-elseif-else 2`] = `
"[IF]
	Hi, [VALUE]?!
[ELSEIF]
	[VALUE]
[ELSE]
	[CALL]
[/IF]
"
`;

exports[`if-elseif-notdef-else 1`] = `
"<!-- IF TestServer && Eq(GET_x,1) -->
	Hi, ##UserName##?
<!-- ELSEIFNOT TestServer -->
	##HOST##
<!-- ELSEIFDEF GET_id -->
	##GET_id##
<!-- ELSEIFNOTDEF GET_id -->
	##EMAIL##
<!-- ELSE -->
	##JsonEncode(GET_id)##
<!-- /IF -->
"
`;

exports[`if-elseif-notdef-else 2`] = `
"[IF]
	Hi, [VALUE]?
[ELSEIFNOT]
	[VALUE]
[ELSEIFDEF]
	[VALUE]
[ELSEIFNOTDEF]
	[VALUE]
[ELSE]
	[CALL]
[/IF]
"
`;

exports[`if-ending 1`] = `
"<!-- IF ActiveUser -->
	Hi, ##GET_name##!
<!-- /IF ActiveUser -->
"
`;

exports[`if-ending 2`] = `
"[IF]
	Hi, [VALUE]!
[/IF]
"
`;

exports[`ifdef 1`] = `
"<!-- IFDEF GET_name -->
	Hi, ##GET_name##
<!-- /IFDEF -->
"
`;

exports[`ifdef 2`] = `
"[IFDEF]
	Hi, [VALUE]
[/IFDEF]
"
`;

exports[`ifnot 1`] = `
"<!-- IFNOT GET_name -->Hi, ##GET_name##<!-- /IFNOT -->
"
`;

exports[`ifnot 2`] = `
"[IFNOT]Hi, [VALUE][/IFNOT]
"
`;

exports[`ifnotdef 1`] = `
"<!-- IFNOTDEF GET_name -->Hi, ##GET_name##<!-- /IFNOTDEF -->
"
`;

exports[`ifnotdef 2`] = `
"[IFNOTDEF]Hi, [VALUE][/IFNOTDEF]
"
`;

exports[`include 1`] = `
"<!-- INCLUDE ./foo.html -->
"
`;

exports[`include 2`] = `
"[INCLUDE]
"
`;

exports[`text 1`] = `
"Hi, #user-name#!
"
`;

exports[`text 2`] = `
"Hi, #user-name#!
"
`;

exports[`value 1`] = `
"Hi, ##UserName##!
"
`;

exports[`value 2`] = `
"Hi, [VALUE]!
"
`;
